### ISSUE TITLE
MethodSpec and TypeSpec add methods to accept Iterator

### ISSUE THREAD

**Issue by 0x3333** [square/javapoet#497]
Most `*Spec` methods, like MethodSpec, TypeSpec, etc, accept `Iterable<?>` in add methods. e.g:

[TypeSpec - `public Builder addSuperinterfaces(Iterable<? extends TypeName> superinterfaces)`](https://github.com/square/javapoet/blob/master/src/main/java/com/squareup/javapoet/TypeSpec.java#L448)
[MethodSpec - `public Builder addExceptions(Iterable<? extends TypeName> exceptions)`](https://github.com/square/javapoet/blob/master/src/main/java/com/squareup/javapoet/MethodSpec.java#L392)

Stream in Java 8 cannot be used inline, like:

```java
MethodSpec.Builder method = MethodSpec.methodBuilder(elMethodName)
        .addExceptions(
                thrownTypes.stream().map(
                        typeMirror -> TypeName.get(typeMirror)).iterator()
                ));
```

**Comment by yevhen13** [comment]
```
MethodSpec.Builder method = MethodSpec.methodBuilder(elMethodName)
        .addExceptions(
                thrownTypes.stream().map(
                        typeMirror -> TypeName.get(typeMirror)
                )::iterator
        );
```
?

**Comment by kenzierocks** [comment]
That may not be safe, since `Iterable.iterator` could not be used twice. It should probably work as a workaround for now, though.

**Comment by 0x3333** [comment]
@yevhen13 this is not a solution. Its a workaround, iterable is different from iterator. Iterable provides iterator, iterator can't be reused.

**Comment by yevhen13** [comment]
> Its a workaround

Well, yes it is... _sadly_.

Though, I think an API method with iterator (or stream) argument isn't really all that useful in practice, and I haven't seen this pattern for a while. By the way, is it used somewhere in the Java API (or any other [popular] java libraries)?

> That may not be safe, since Iterable.iterator could not be used twice.

Why would you need to iterate it twice (in these methods)?

'Single iteration' behavior can (?) be documented in the methods' javadocs.
Otherwise,
- collect an intermediate collection; or
- create an iterable 'stream factory' - that will perform full chain of stream operations on original collection for each iteration.

**Comment by JakeWharton** [comment]
Iterable parameters are very common. With Java 9's list, map, and set factories the need for APIs to accept both varargs and iterables will diminish (favoring the latter).

**Comment by yevhen13** [comment]
@JakeWharton 

> Iterable parameters are very common.

Agreed. Then what do you think about Iterator(s)/Stream(s) as parameters?

> With Java 9's list, map, and set factories the need for APIs to accept both varargs and iterables will diminish (favoring the latter).

Yeah. And I just hope that the need for arrays (especially primitive) will someday diminish too.

**Comment by jbgi** [comment]
also a method modifying its arguments is generally considered bad practice.

**Comment by JakeWharton** [comment]
> Then what do you think about Iterator(s)/Stream(s) as parameters?

Oh I missed that's what the issue was actually about. I see no reason to support this. Either collect to a list or invert the dependency and forEach items into the builder.

**Comment by 0x3333** [comment]
> Though, I think an API method with iterator (or stream) argument isn't really all that useful in practice, and I haven't seen this pattern for a while. By the way, is it used somewhere in the Java API (or any other [popular or not] java libraries)?

@yevhen13 a simple github search gave me almost 10mi usages, java.util.List gave me 30mi.
33% of usage in comparison to a API that is since 1.2. To me it sounds pretty reasonable.

> Why would you need to iterate it twice (in these methods)?
> 
> 'Single iteration' behavior can (?) be documented in the methods' javadocs.

I don't need to, in fact, the javapoet doesn't, it's just not semantically.

> collect an intermediate collection; or
> create an iterable 'stream factory' - that will perform full chain of stream operations on original collection for each iteration; or
> use TypeSpec.addSuperinterface / MethodSpec.addException with outer loop / forEach.

This would just trash the benefits from the Stream API.

**Comment by 0x3333** [comment]
@jbgi I don't want to modify, I want to add.

**Comment by 0x3333** [comment]
@JakeWharton don't you believe that this would make usage less verbose?

Usually, code that use javapoet is dense, it would make it easier to read, IMHO.

Anyways, I collected everything that uses stream to lists and called the methods. It became clearer.

**Comment by jbgi** [comment]
@0x3333 I meant the stream/iterator state is modified by iterating on them.

**Comment by 0x3333** [comment]
@jbgi agreed. But in the case of a Iterator, you can't use without modifying...
